packages/skott/src/skott.ts

```diff
@@ -485,12 +528,133 @@ export class Skott<T> {
     };
   }
 
+  private getValidGroup(nodePath: string) {
+    if (!this.config.groupBy) {
+      throw new Error(
+        "Grouped graph can't be built without a groupBy function in the configuration"
+      );
+    }
+
+    const result = this.config.groupBy(nodePath);
+
+    if (typeof result === "string" || !result) {
+      return result;
+    }
+
+    throw new Error(
+      `groupBy function must return a string or undefined, but returned "${result}" (for "${nodePath}")`
+    );
+  }
+
+  private addNodeToGroupedGraph(node: SkottNode<T>) {
+    const group = this.getValidGroup(node.id);
+
+    if (group) {
+      /**
+       * Typecast is ok here:
+       * current typings are expecting SkottNodeBody + other arbitary values if available, but currently TS is not happy with it.
+       *
+       * The SkottNode type should be described in some other way to properly support that, but SkottNodeBody is avaiable at the runtime anyway.
+       */
+      const nodeBody = node.body as SkottNodeBody;
+
+      if (this.#groupedGraph!.hasVertex(group)) {
+        /**
+         * Group vertex already exists, we need to add up:
+         * - the file to the group
+         * - the size of the new node
+         * - built-in and third-party dependencies
+         */
+        this.#groupedGraph!.mergeVertexBody(group, (groupBody) => {
+          if (groupBody.files.includes(node.id)) return;
+
+          groupBody.files.push(node.id);
+
+          groupBody.size += nodeBody.size;
+
+          nodeBody.thirdPartyDependencies.forEach((dep) => {
+            if (!groupBody.thirdPartyDependencies.includes(dep)) {
+              groupBody.thirdPartyDependencies.push(dep);
+            }
+          });
+
+          nodeBody.builtinDependencies.forEach((dep) => {
+            if (!groupBody.builtinDependencies.includes(dep)) {
+              groupBody.builtinDependencies.push(dep);
+            }
+          });
+        });
+      } else {
+        /**
+         * Group vertex does not exist yet, we need to create it
+         *
+         * Initial size is the size of the first node, as for all kinds of dependencies
+         */
+        this.#groupedGraph!.addVertex({
+          id: group,
+          adjacentTo: [],
+          body: {
+            size: nodeBody.size,
+            files: [node.id],
+            thirdPartyDependencies: [...nodeBody.thirdPartyDependencies],
+            builtinDependencies: [...nodeBody.builtinDependencies]
+          }
+        });
+      }
+    }
+  }
+
+  private buildGroupedGraph(
+    projectStructure: Record<string, SkottNode<T>>
+  ): DiGraph<SkottNode<{ files: string[] }>> {
+    this.#groupedGraph = new DiGraph<SkottNode<{ files: string[] }>>();
+
+    for (const node of Object.values(projectStructure)) {
+      this.addNodeToGroupedGraph(node);
+
+      /**
+       * Update edges
+       */
+      node.adjacentTo.forEach((adjacentNode) => {
```

Comment:
```suggestion
      node.adjacentTo.forEach((adjacentNodeId) => {
```

`adjacentTo` is only a list of node ids so I to avoid any confusion there I would suggest you to rename the variable in the callback